core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -1117,6 +1160,136 @@ object Chunk {
     override def toArray[A1 >: A](n: Int, dest: Array[A1]): Unit = { val _ = vector.copyToArray(dest, n, length) }
   }
 
+  private[zio] object Buff {
+    final def arrayBacked[A](buffer: Buffer): Option[Chunk[A]] =
+      if (!buffer.hasArray()) None
+      else {
+        val pos = buffer.position()
+        val lim = buffer.limit()
+        val cap = buffer.capacity()
+        val arr = buffer.array().asInstanceOf[Array[A]]
+        val off = buffer.arrayOffset()
+
+        val chunk = if (off == 0 && pos == 0 && cap == arr.length && lim == cap) {
+          // best case, reuse the underlying array
+          fromArray(arr)
+        } else {
+          implicit val classTag: ClassTag[A] = ClassTag(arr.getClass.getComponentType)
+          val len                            = lim - pos
+          val dest                           = Array.ofDim[A](len)
+          Array.copy(arr, off + pos, dest, 0, len)
+          fromArray(dest)
+        }
+
+        Some(chunk.asInstanceOf[Chunk[A]])
+      }
+
+    final def manualByteBuffer(buffer: ByteBuffer): Chunk[Byte] = {
+      val pos = buffer.position()
+      val lim = buffer.limit()
+      val len = lim - pos
+      val dst = Array.ofDim[Byte](len)
+      var i   = pos
+      var j   = 0
+      while (i < lim) {
+        dst(j) = buffer.get(i)
```

Comment:
This is really slow. Let's use `Buffer#get(arr)` instead.

Reply:
The reason I implemented this is because I was not sure if we are allowed to use the `mark()` and bulk `get` methods.

Reply:
To me, `mark()`ing a Buffer is conceptually the same as mutating it.

Reply:
I think given how buffers work I think the 'expected api' would be:
```scala
buf.mark()
val chunk = Chunk.manualByteBuffer(buf)
buf.reset()
```
This would of course make manualByteBuffer non-rt, but I think its fine as we are in mutable land anyway here. Maybe call it unsafeFromBuffer ?

Reply:
Marking and resetting is a no-op (or should be) so I wouldn’t say it makes this method non-RT.

Reply:
Err that is unless mark was already used before.

Reply:
Yes, that's my point. Also, `reset()` doesn't remove the mark.

Reply:
How about:
```scala
val pos = buffer.position()
buffer.get(dest)
buffer.position(pos)
```
?

Reply:
Yup! That’s better.

Reply:
+1. This is nice

Reply:
> Marking and resetting is a no-op (or should be) so I wouldn’t say it makes this method non-RT.

One point is overwriting the mark. Another is that I was thinking of leaving this to the user completely. So I think we could change mark and position at will, and document that if desired the user has to call mark/reset

Reply:
Are we ok with the current (new) implementation?